### PR TITLE
spectral-extraction: support for optimal extraction

### PR DESCRIPTION
**Pull request by kecnry**
### Description

This pull request adds support for optimal extraction (in addition to the existing boxcar extraction) to the spectral extraction plugin in specviz2d.

- [x]  requires https://github.com/astropy/specreduce/pull/118 to be merged/released first (otherwise will return a spectrum of all zeros).

**TODO**:
- [x] merge, release, pin specreduce bugfix
- [x] determine whether to map spectrum1d uncertainties to variances

### Checklist for package maintainer(s)

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [x] Are two approvals required? Branch protection rule does not check for the second approval. If a second approval is not necessary, please apply the `trivial` label.
- [x] Do the proposed changes actually accomplish desired goals? Also manually run the affected example notebooks, if necessary.
- [x] Do the proposed changes follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Are tests added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Are docs added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [x] Did the CI pass? If not, are the failures related?
- [x] Is a change log needed? If yes, is it added to `CHANGES.rst`?
- [x] Is a milestone set?
- [ ] After merge, any internal documentations need updating (e.g., JIRA, Innerspace)?

### CODE REVIEW

File: jdaviz/configs/specviz2d/plugins/spectral_extraction/spectral_extraction.py
Comment (kecnry) [reply jdaviz/configs/specviz2d/plugins/spectral_extraction/spectral_extraction.py:16]: Orrrrr possibly if they're just being used as weights, we might not care if they're actually variances and can just pass whatever is attached to the spectrum1d object.

Comment (pllim) [reply jdaviz/configs/specviz2d/plugins/spectral_extraction/spectral_extraction.py:16]: Maybe @rosteen can advise. 🙏

Comment (rosteen) [reply jdaviz/configs/specviz2d/plugins/spectral_extraction/spectral_extraction.py:16]: I would be fine making that function public in `specutils`, but you may not want to wait for that to merge this.

Comment (kecnry) [reply jdaviz/configs/specviz2d/plugins/spectral_extraction/spectral_extraction.py:16]: if we wanted to do that, would that change then break jdaviz if this makes it in (since specutils probably wouldn't deprecate the hidden function)?

Comment (rosteen) [reply jdaviz/configs/specviz2d/plugins/spectral_extraction/spectral_extraction.py:16]: I guess the timing could be a little awkward on that...hmm. This is where a branching strategy to enable easier bugfix releases would come in handy 😅

Comment (kecnry) [reply jdaviz/configs/specviz2d/plugins/spectral_extraction/spectral_extraction.py:16]: would could hold off on this PR or move the translation logic into either jdaviz or specreduce.  @ojustino - any thoughts on how best to handle passing variances to specreduce's optimal extraction from specutils/jdaviz?

Comment (ojustino) [reply jdaviz/configs/specviz2d/plugins/spectral_extraction/spectral_extraction.py:16]: I wish I had known this function would while I was working on `OptimalExtract`; I found my way to [almost the same code](https://github.com/astropy/specreduce/blob/6a546e/specreduce/extract.py#L302:L321) over in `specreduce`. I think it's simple enough that implementing it on the `jdaviz` side shouldn't add too much overhead if we need it now, but big picture, I support having `_convert_uncertainty()` as a public function and using it instead.

File: jdaviz/configs/specviz2d/plugins/spectral_extraction/spectral_extraction.py
Comment (ojustino) [inline jdaviz/configs/specviz2d/plugins/spectral_extraction/spectral_extraction.py:535]: Is your choice of "Optimal" versus "Horne" in the dropdown intentional? I don't have a preference, but I wonder if others do.

File: jdaviz/configs/specviz2d/plugins/spectral_extraction/spectral_extraction.py
Comment (kecnry) [reply jdaviz/configs/specviz2d/plugins/spectral_extraction/spectral_extraction.py:16]: ok, so for now what do you think of jdaviz passing an NDData object with the uncertainty passed in its own native type to let specreduce handle the conversion?

Comment (kecnry) [reply jdaviz/configs/specviz2d/plugins/spectral_extraction/spectral_extraction.py]: @rosteen @ojustino - let me know what you think of the latest commit which refactors the logic to pass an NDData object, assumes the uncertainty is standard deviation if not defined in the spectrum1d object, and defaults to a variance of all 1s if no uncertainty is attached.

File: jdaviz/configs/specviz2d/plugins/spectral_extraction/spectral_extraction.py
Comment (kecnry) [reply jdaviz/configs/specviz2d/plugins/spectral_extraction/spectral_extraction.py]: Changed to "Horne" and added links to the specreduce API docs (which should be expanded to provide more details on the individual implementations).

File: jdaviz/configs/specviz2d/plugins/spectral_extraction/spectral_extraction.py
Comment (ojustino) [inline jdaviz/configs/specviz2d/plugins/spectral_extraction/spectral_extraction.py:631]: When are you thinking the `if not` clause will be triggered? I'm not aware of a case where the `uncertainty` attribute for either `CCDData` or `NDData` lacks an `uncertainty_type` attribute of its own.

Also, while it's allowed, specifying a `StdDevUncertainty`-type `uncertainty` currently triggers a warning from `HorneExtract` because the assumption is that people will want to pass variances instead.

Comment (ojustino) [reply jdaviz/configs/specviz2d/plugins/spectral_extraction/spectral_extraction.py:631]: My question on `StdDevUncertainty` above still applies, but I think you may need to change the `if not` statement. For example, I get an `uncertainty_type` of `'unknown'` (not `None`) from the following:
```
NDData(np.arange(10), uncertainty=np.ones(10)).uncertainty.uncertainty_type
```

Comment (kecnry) [reply jdaviz/configs/specviz2d/plugins/spectral_extraction/spectral_extraction.py:631]: `inp_sp2d` is a specutils Spectrum1D object, and I believe attaching an uncertainty is optional, and could be just an array (in which case `inp_sp2d.uncertainty` is an array which does not have the `uncertainty_type` attribute`).  So we need to decide fallbacks for at least two cases:
* when no uncertainty exists on the Spectrum1D object.  Specreduce would fallback on variances of all ones if not passing NDData, so I figured we should do the same here.
* when the uncertainty on the Spectrum1D object is an array (without a type).  This one is tricky.  You're right that specreduce will raise a warning if we assume `StdDevUncertainty`, but maybe a warning in this case is a good thing?  

And possibly your case, if that is possible to produce in a Spectrum1D object.  Hopefullly @rosteen can help give some insight from the specutils side of things.

Comment (rosteen) [reply jdaviz/configs/specviz2d/plugins/spectral_extraction/spectral_extraction.py:631]: The uncertainty on a `Spectrum1D` is supposed to be a `NDUncertainty` object, not a bare array, but looking at the code I don't think there's actually a check to make sure the input uncertainty is an `NDUncertainty`. We should probably enforce that in `specutils`.

Comment (rosteen) [reply jdaviz/configs/specviz2d/plugins/spectral_extraction/spectral_extraction.py:631]: I was debating whether it would be best to add ` or uncert.uncertainty_type == "unknown"` to the `hasattr` check, but honestly I think it's ok to have an error raised rather than making an assumption on the user's behalf.

**Comment by kecnry** [comment]
@rosteen - the button should disable itself and display a message if it expects an error from specreduce.  I was able to reproduce this, but only on a stale notebook.  Can you refresh and see if it still happens?

**Review by rosteen — APPROVED** [review]
That fixed the problem I was seeing with the `Extract` button. Approving, but the changelog needs to be updated before merging.

**Comment by kecnry** [comment]
coverage was passing before rebasing on main, btw, but can add more tests if anyone thinks any of the uncovered lines needs to be covered.

File: jdaviz/configs/specviz2d/plugins/spectral_extraction/spectral_extraction.py
Comment (ojustino) [reply jdaviz/configs/specviz2d/plugins/spectral_extraction/spectral_extraction.py:631]: OK, I think the name/type combination for `inp_sp2d` confused me. I now understand it as a `Spectrum1D` object whose `flux` attribute is a 2D array of the image. Is that right?

Your second case is triggered by a scenario that isn't prohibited, but that we also don't want to encourage – using a bare array for a `Spectrum1D` object's `uncertainty` attribute. Are you saying that since `specutils` currently doesn't prohibit that scenario, you're using `StdDevUncertainty` on purpose to trigger the warning from `HorneExtract` as an indirect way to make the user think twice about providing that kind of input?

If so, I think it makes more sense to always assume `VarianceUncertainty` on the Jdaviz side for unclear uncertainties and to be more explicit about what we expect from the `Spectrum1D` object imported by the user. I guess this scenario becomes moot if @rosteen is able to enforce that all uncertainties must be in the `NDUncertainty` family, but there may be a reason why it's not that way.

Comment (kecnry) [reply jdaviz/configs/specviz2d/plugins/spectral_extraction/spectral_extraction.py:631]: > I now understand it as a Spectrum1D object whose flux attribute is a 2D array of the image. Is that right?

Yes, that's right.

> Your second case is triggered by a scenario that isn't prohibited, but that we also don't want to encourage – using a bare array for a Spectrum1D object's uncertainty attribute. 

I agree - this isn't a use we should encourage, but since specutils _allows_ it, we have to handle it somehow (or catch it earlier in specutils or jdaviz and raise an error).

> Are you saying that since specutils currently doesn't prohibit that scenario, you're using StdDevUncertainty on purpose to trigger the warning from HorneExtract as an indirect way to make the user think twice about providing that kind of input?

A little bit of that but also somewhat assuming that people generally default to standard deviation (could be field-dependent though).  I would be a bit scared to default to something that doesn't raise a warning though (unless we raise our own warning).

> I guess this scenario becomes moot if @rosteen is able to enforce that all uncertainties must be in the NDUncertainty family, but there may be a reason why it's not that way.

That's true - @rosteen - if you think this would be a useful to have on specutils, should we create an issue/ticket there and then make a note to come back here and remove this if-statement once able?

Comment (rosteen) [reply jdaviz/configs/specviz2d/plugins/spectral_extraction/spectral_extraction.py:631]: I think it should go into `specutils`, with the caveat that getting it in there might take anywhere between two weeks and three months 😬 

There's some old discussion of this [here](https://github.com/astropy/specutils/issues/529), I'd be happy to make this an upcoming priority in `specutils`. But that doesn't help us immediately for this PR.

Comment (ojustino) [reply jdaviz/configs/specviz2d/plugins/spectral_extraction/spectral_extraction.py:631]: In that case, I think it would be good to issue a warning or notification about the standard deviation assumption when a `Spectrum1D` object with a numpy `array` uncertainty attribute is imported instead of during the export process. I'm also curious about whether standard deviation or variance is the best assumption to make in that case, but I don't have the context to know.

Comment (kecnry) [reply jdaviz/configs/specviz2d/plugins/spectral_extraction/spectral_extraction.py:640]: I added an in-line plugin warning (below the dropdown where Horne is selected).  You can trigger it manually (for testing how it looks) with:

```
pext = specviz2d.app.get_tray_item_from_name('spectral-extraction')
pext.ext_uncert_warn = True
```

image

**Comment by kecnry** [comment]
this is now failing CI due to the backgrounds-overlap error (which should be fixed by specreduce#125, although I have no idea why this only started failing with the latest commit which shouldn't have affected any of the internal logic at all... is it possible the remote data used in that test has changed and now triggers the background error but didn't before? 🤔 )

File: jdaviz/configs/specviz2d/plugins/spectral_extraction/spectral_extraction.py
Comment (ojustino) [inline jdaviz/configs/specviz2d/plugins/spectral_extraction/spectral_extraction.py:405]: To cover the numpy array uncertainty case we've been discussing, I believe you need something like `and inp_sp2d.uncertainty != 'unknown'` at the end of this statement. I tested by setting `inp_sp2d.uncertainty = np.random.rand(*inp_sp2d.flux.shape)` just above it, and the warning doesn't display in Specviz2D with this check as it's currently constructed.

Comment (kecnry) [reply jdaviz/configs/specviz2d/plugins/spectral_extraction/spectral_extraction.py]: ah, it seems specutils is wrapping the bare array in an `UnknownUncertainty` object.  I changed the logic to just check for that explicitly now.

Comment (ojustino) [reply jdaviz/configs/specviz2d/plugins/spectral_extraction/spectral_extraction.py]: That's a better way to go about it, and it works for me now.

File: jdaviz/configs/specviz2d/plugins/spectral_extraction/spectral_extraction.py
Comment (ojustino) [reply jdaviz/configs/specviz2d/plugins/spectral_extraction/spectral_extraction.py:640]: Could we include short instructions in the warning about how to remedy the issue if using standard deviations not OK? Since the `Spectrum1D` object they imported (`s1d`) should already have content in its `uncertainty` attribute, I believe they would just need to wrap `s1d.uncertainty.array` in a differently-typed `NDUncertainty` class, assign that as `s1d`'s uncertainty, and reimport `s1d` into Specviz2D.

**Comment by pllim** [comment]
I don't grok this stuff so if @ojustino can give the second approval, that would be great. Thanks!

File: CHANGES.rst
Comment (pllim) [inline CHANGES.rst:21]:
```suggestion
Specviz2d
```

File: jdaviz/configs/specviz2d/plugins/spectral_extraction/spectral_extraction.py
Comment (ojustino) [reply jdaviz/configs/specviz2d/plugins/spectral_extraction/spectral_extraction.py:640]: If the plugin gets much outside use, I don't think this scenario will be so rare. From past discussions with product owners about `HorneExtract`, it seems like many scientific users of objects like `CCDData`, `NDData`, `Spectrum1D`, etc. don't use the `NDUncertainty` family if they can get away with it because they're more used to lists and numpy arrays. Since @rosteen cautioned us against expecting the fix in `specutils` soon, I think it's good to have something to which we can point in the meantime.

I see what you're saying about the difficulty of knowing how the `Spectrum1D` object was loaded. Maybe this would provide more context: "Imported 2D image's uncertainties have unclear units; assuming standard deviation."

We could also add a short sentence in the part of `docs/specviz2d/import_data.rst` that discusses `Spectrum1D` saying that uncertainties are assumed to be standard deviations unless otherwise specified.

Comment (kecnry) [reply jdaviz/configs/specviz2d/plugins/spectral_extraction/spectral_extraction.py:640]: I changed the warning text to: "WARNING: uncertainties on input 2D spectrum have unclear units; assuming standard deviation."

And added a note in the spectral extraction plugin docs (since for now that is the only place this assumption is made), saying: "Note: Horne extraction uses uncertainties on the input 2D spectrum.  If the spectrum uncertainties are not explicitly assigned a type, they are assumed to be standard deviation uncertainties."

File: docs/specviz2d/plugins.rst
Comment (ojustino) [inline docs/specviz2d/plugins.rst:134]:
```suggestion
.. note::

    Horne extraction uses uncertainties on the input 2D spectrum. If the
    spectrum uncertainties are not explicitly assigned a type, they are assumed
    to be standard deviation uncertainties.
```

File: jdaviz/configs/specviz2d/plugins/spectral_extraction/spectral_extraction.py
Comment (ojustino) [reply jdaviz/configs/specviz2d/plugins/spectral_extraction/spectral_extraction.py:640]: I think these are fair changes. I just suggested using the Sphinx's "note" directive for consistency with other pages, but the content looks good to me.

**Review by ojustino — APPROVED** [review]

**Pull request merged**
